gramps/gen/filters/optimizer.py

```diff
@@ -73,11 +76,29 @@ def compute_potential_handles_for_rule(
         self,
         rule: Rule,
     ) -> Tuple[Set[PrimaryObjectHandle] | None, Set[PrimaryObjectHandle] | None]:
-        """ """
+        """
+        Find the the handles for a particular rule.
+        """
+        # Has to be of the appropriate type
         if hasattr(rule, "selected_handles"):
             return (rule.selected_handles, None)
         if hasattr(rule, "find_filter"):
             filter = rule.find_filter()
-            if filter:
+            if filter and self.is_same_namespace(filter):
                 return self.compute_potential_handles_for_filter(filter)
         return (None, None)
+
+    def is_same_namespace(self, filter):
+        """
+        Determine if the given filter is in the 'same namespace' as the top-level filter.
+
+        In this context, 'same namespace' means that both filters are instances of the same class,
+        as determined by comparing their class names.
+
+        Parameters:
+            filter: The filter object to compare against the top-level filter.
+
+        Returns:
+            True if both filters have the same class name (i.e., are of the same type), False otherwise.
+        """
+        return type(self.top_level_filter) == type(filter)
```

Comment:
I think this is safe, but could be too cautious. As an example, `DeferredFamilyFilter` subclasses `GenericFamilyFilter`. Therefore their types are different and `is_same_namespace` will return false. [this is from reading the code, I've not constructed a test case to see if it is actually possible to contruct a rule using a filter of each of these types]

We really want to check the type of the objects returned by the filter, perhaps a property can be added to `GenericFilter`, initialised by the the `Generic*Filter` classes. This property can then be used in `is_same_namespace`